libclamav_rust/src/fuzzy_hash.rs

```diff
@@ -110,212 +118,205 @@ pub struct FuzzyHashMeta {
     hexsig: *const c_char,
     lsig_id: u32,
     subsig_id: u32,
     err: *mut *mut FRSError,
 ) -> bool {
     let hexsig = validate_str_param!(hexsig);
 
-    let mut hashmap = unsafe {
-        ManuallyDrop::new(Box::from_raw(
-            fuzzy_hashmap as *mut HashMap<FuzzyHash, Vec<FuzzyHashMeta>>,
-        ))
-    };
-
-    let load_result = fuzzy_hash_load_subsignature(&mut hashmap, hexsig, lsig_id, subsig_id);
+    let mut hashmap = ManuallyDrop::new(Box::from_raw(fuzzy_hashmap as *mut FuzzyHashMap));
 
-    unsafe { frs_result!(result_in = load_result, err = err) }
-}
-
-/// Load a fuzzy hash subsignature
-/// Parse a fuzzy hash logical sig subsignature.
-/// Add the fuzzy hash to the matcher so it can be matched.
-pub fn fuzzy_hash_load_subsignature(
-    hashmap: &mut Box<HashMap<FuzzyHash, Vec<FuzzyHashMeta>>>,
-    hexsig: &str,
-    lsig_id: u32,
-    subsig_id: u32,
-) -> Result<(), FuzzyHashError> {
-    let mut hexsig_split = hexsig.split('#');
-
-    let algorithm = match hexsig_split.next() {
-        Some(x) => x,
-        None => return Err(FuzzyHashError::Format),
-    };
-
-    let hash = match hexsig_split.next() {
-        Some(x) => x,
-        None => return Err(FuzzyHashError::Format),
-    };
-
-    let distance: u32 = match hexsig_split.next() {
-        Some(x) => match x.parse::<u32>() {
-            Ok(n) => n,
-            Err(_) => {
-                return Err(FuzzyHashError::FormatHammingDistance(x.to_string()));
-            }
-        },
-        None => 0,
-    };
-
-    // TODO: Support non-zero distance
-    if distance != 0 {
-        error!(
-            "Non-zero hamming distances for image fuzzy hashes are not supported in this version."
-        );
-        return Err(FuzzyHashError::InvalidHammingDistance(distance));
-    }
-
-    match algorithm {
-        "fuzzy_img" => {
-            // Convert the hash string to an image fuzzy hash bytes struct
-            let image_fuzzy_hash = hash
-                .try_into()
-                .map_err(|e| FuzzyHashError::FormatHashBytes(format!("{}: {}", e, hash)))?;
-
-            let fuzzy_hash = FuzzyHash::Image(image_fuzzy_hash);
-
-            let meta: FuzzyHashMeta = FuzzyHashMeta {
-                lsigid: lsig_id,
-                subsigid: subsig_id,
-                #[cfg(feature = "not_ready")]
-                hamming_distance: distance,
-            };
-
-            // If the hash key does not exist in the hashmap, insert an empty vec.
-            // Then add the current meta struct to the entry.
-            hashmap
-                .entry(fuzzy_hash)
-                .or_insert_with(Vec::new)
-                .push(meta);
-
-            Ok(())
-        }
-        _ => {
-            error!("Unknown fuzzy hash algorithm: {}", algorithm);
-            Err(FuzzyHashError::UnknownAlgorithm(algorithm.to_string()))
-        }
-    }
+    frs_call!(
+        err = err,
+        hashmap.load_subsignature(hexsig, lsig_id, subsig_id)
+    )
```

Comment:
Having the macro params separate from the function params does feel much better. I like it.